### PR TITLE
[hybrid planning] Add action abortion and test; improve the existing test

### PR DESCRIPTION
**Pull request by AndyZe**
- Add the ability to cancel a hybrid planning action request
- Remove the existing Python integration test. It was very basic; it only checked that certain nodes did not die
- Add a C++ integration test that checks for action completion
- Add a C++ integration test for the new cancelation feature
- Do not continue publishing the same trajectory over and over if the global planner needs to replan
- Fix the clearing of robot trajectory if the trajectory is invalid (use `.clear()` instead of `.empty()`)
- Ensure the initial state of the robot is set properly

@sjahr

### CODE REVIEW

**Comment by AndyZe** [comment]
CI is failing due to something related to controller loading. Any ideas, @destogl or @jackcenter ?

```
 [component_container_mt-1] [FATAL] [1641768093.218988192] [moveit_ros.trajectory_execution_manager]: Exception while loading controller manager 'moveit_simple_controller_manager/MoveItSimpleControllerManager': According to the loaded plugin descriptions the class moveit_simple_controller_manager/MoveItSimpleControllerManager with base class type moveit_controller_manager::MoveItControllerManager does not exist. Declared types are 
Error: mponent_container_mt-1] [ERROR] [1641768093.219830987] [moveit_ros.trajectory_execution_manager]: Failed to reload controllers: `controller_manager_` does not exist.
```

**Comment by tylerjw** [comment]
> 

My guess is you are missing a run_depend and it works locally because you have sourced your workspace you are running the test in (which adds all the workspace libraries to your path regardless of what is specified in the package.xml).

**Comment by tylerjw** [comment]
> My guess is you are missing a run_depend and it works locally because you have sourced your workspace you are running the test in (which adds all the workspace libraries to your path regardless of what is specified in the package.xml).  The evidence is that plugin loader doesn't find any types that match that base type string.

**Comment by AndyZe** [comment]
This PR now prevents behavior like this:

local_planner_stuck

File: moveit_ros/hybrid_planning/test/test_basic_integration.cpp
Comment (tylerjw) [inline moveit_ros/hybrid_planning/test/test_basic_integration.cpp:95]: Passing tf2_ros::Buffer to PlanningSceneMonitor's constructor is deprecated

File: moveit_ros/hybrid_planning/test/test_basic_integration.cpp
Comment (tylerjw) [inline moveit_ros/hybrid_planning/test/test_basic_integration.cpp:104]: is this needed for your test in some way?

File: moveit_ros/hybrid_planning/test/test_basic_integration.cpp
Comment (tylerjw) [inline moveit_ros/hybrid_planning/test/test_basic_integration.cpp:145]: use inline lambda initialization, there seems to be no reason this should be mutable

File: moveit_ros/hybrid_planning/test/test_basic_integration.cpp
Comment (tylerjw) [inline moveit_ros/hybrid_planning/test/test_basic_integration.cpp:149]: again, lambda initalization

File: moveit_ros/hybrid_planning/test/test_basic_integration.cpp
Comment (tylerjw) [inline moveit_ros/hybrid_planning/test/test_basic_integration.cpp:199]: Does this need to be overriden?

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h
Comment (tylerjw) [inline moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h:124]: Is this being used to communicate between concurrent sections of code?  If so, please use a lock-free queue or other message-passing technique instead of shared memory.

Here is a lock-free queue: https://www.boost.org/doc/libs/1_76_0/doc/html/boost/lockfree/spsc_queue.html

For reasoning read this blog post: http://henrikeichenhardt.blogspot.com/2013/06/why-shared-mutable-state-is-root-of-all.html

For examples from other languages see this section from the Rust Book: https://doc.rust-lang.org/book/ch16-02-message-passing.html and this section from the Golang documentation: https://go.dev/doc/effective_go#concurrency

For an alternate solution, you could use strong typing and have the code divided between behavior for executing and not executing and change a type that determines the action being taken.  Then you could use RAII for cleaning up and transitioning between your two states.

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp
Comment (tylerjw) [inline moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp:47]: I see we never got rid of the ManagerManager name or the inheriting from rclcpp::Node.   :disappointed:

File: moveit_ros/hybrid_planning/local_planner/local_planner_component/src/local_planner_component.cpp
Comment (tylerjw) [inline moveit_ros/hybrid_planning/local_planner/local_planner_component/src/local_planner_component.cpp:164]: This looks like communication across threads with shared memory also.  See my previous comment on that.

File: moveit_ros/hybrid_planning/test/test_basic_integration.cpp
Comment (AndyZe) [reply moveit_ros/hybrid_planning/test/test_basic_integration.cpp:60]: I guess not. Done.

File: moveit_ros/hybrid_planning/test/test_basic_integration.cpp
Comment (AndyZe) [reply moveit_ros/hybrid_planning/test/test_basic_integration.cpp:62]: I think you might have read that wrong? It's node subclassing a node.

Comment (tylerjw) [reply moveit_ros/hybrid_planning/test/test_basic_integration.cpp:62]: you are right, I did read this wrong.

File: moveit_ros/hybrid_planning/test/test_basic_integration.cpp
Comment (tylerjw) [inline moveit_ros/hybrid_planning/test/test_basic_integration.cpp:60]: I meant delete the empty function, not remove the override keyword.

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp
Comment (tylerjw) [inline moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp:168]: If you want to use the function instead of an inline lambda you can call the function from within the lambda and the compiler will optimize away the lambda.  You can also use `auto` for the type of parameter to the lambda and it should deduce the type if you use it to call another function.

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp
Comment (tylerjw) [inline moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp:185]: This looks like another thing that could be const if initialized in a lambda.

reference: http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-lambda-init

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp
Comment (tylerjw) [inline moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp:187]: If possible use stream concatination instead of %s.  %s is a known security hole.

File: moveit_ros/hybrid_planning/test/test_basic_integration.cpp
Comment (AndyZe) [reply moveit_ros/hybrid_planning/test/test_basic_integration.cpp:105]: nope, just copied from the demo

File: moveit_ros/hybrid_planning/test/test_basic_integration.cpp
Comment (tylerjw) [inline moveit_ros/hybrid_planning/test/test_basic_integration.cpp:92]: I'm confused, you don't need to pass this parameter.  Why are you just commenting it out?

File: moveit_ros/hybrid_planning/test/test_basic_integration.cpp
Comment (AndyZe) [reply moveit_ros/hybrid_planning/test/test_basic_integration.cpp:145]: what's the point of this?

File: moveit_ros/hybrid_planning/test/test_basic_integration.cpp
Comment (AndyZe) [reply moveit_ros/hybrid_planning/test/test_basic_integration.cpp:92]: Still in the middle of changes. Hang tight.

File: moveit_ros/hybrid_planning/test/test_basic_integration.cpp
Comment (AndyZe) [reply moveit_ros/hybrid_planning/test/test_basic_integration.cpp:145]: ah, to make it const. I see

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp
Comment (AndyZe) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp]: that sounds complicated.

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp
Comment (AndyZe) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp:47]: We talked about it and decided it actually did make some sense in this case. I don't want to dig back through the different PR's to find the conversation right now, though

Comment (tylerjw) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp:47]: I think we decided we were going to fix this.  I don't remember it being good for any reason.  I think I had a PR that fixed it too.  That being said I won't block this change for fixing this.

Comment (tylerjw) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp:47]: I think it only made sense because we didn't have time to fix it at the time you were pushing to merge Hybrid Planner.  I don't remember any other reason but I do expect we will either fix this or document very clearly why this is necessary.

Comment (AndyZe) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp:47]: https://github.com/ros-planning/moveit2/pull/488#discussion_r720463753

Comment (AndyZe) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp:47]: What do you want to call it then? I'll do a quick find/replace.

Comment (tylerjw) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp:47]: The link to the comment didn't work for some reason.  It just opens that giant PR.

I forgot I mentioned the SomethingManager.  I care much less about that than the inheriting from Node.  I don't have a better suggestion for the name at this point but I do know I don't like inheriting from the node object.

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h
Comment (AndyZe) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h]: Don't you need to know how many threads are talking to use spsc queue? I don't know that here. We could probably assume 2 but...

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp
Comment (tylerjw) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp:47]: https://github.com/ros-planning/moveit2/pull/488#discussion_r720408829

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h
Comment (tylerjw) [inline moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h:110]: sadness... raw synchronization primitive and using mutable state

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h
Comment (tylerjw) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h]: yes, you do... you need one producer and one consumer thread or it is not thread safe.  if you have multiple consumers of this you should use a multi-producer multi-consumer queue: https://www.boost.org/doc/libs/1_59_0/doc/html/boost/lockfree/queue.html

Comment (tylerjw) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h]: If you do have multiple consumers you'll need to know how many to make sure you notify each.  It might be better to have a stack of spsc queues that consumers can register themselves into.

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h
Comment (tylerjw) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h:110]: You might find this interesting: https://en.wikipedia.org/wiki/Amdahl%27s_law

There are two reasons to remove shared memory between threads.  

The first is that it is a synchronization point (a point in the software that must be single-threaded) and that the more of those you add the worse our software can scale across multiple processors.  See above link.

The second reason is that they are mutable state.  See the C++ seasoning talk for more reasoning on that.

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h
Comment (AndyZe) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h]: So how many producers and consumers do you think there are? I'm not trying to be difficult here, but that number needs to be known. I think it depends on whether a SingleThreadedExecutor or MultiThreadedExecutor is being used. MultiThreadedExecutor [defaults](https://github.com/ros2/rclcpp/issues/441) to match (num. hardware threads), it looks like. None of that stuff is determined in this header file.

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h
Comment (AndyZe) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h:110]: My understanding is atomics are lock-free since the read or write happens in a single step. So the first bullet point doesn't apply. I'm not sure about the second one.

File: moveit_ros/hybrid_planning/local_planner/local_planner_component/src/local_planner_component.cpp
Comment (AndyZe) [reply moveit_ros/hybrid_planning/local_planner/local_planner_component/src/local_planner_component.cpp:164]: Agree that this is a problem!

Comment (AndyZe) [reply moveit_ros/hybrid_planning/local_planner/local_planner_component/src/local_planner_component.cpp:164]: I can make this variable atomic. That's a lock-free synchronization primitive. I'll push this change; let me know if you can think of a better way (hopefully not ridiculously complicated).

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp
Comment (AndyZe) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp:47]: Made an issue re. inheriting from Node here:  https://github.com/ros-planning/moveit2/issues/1013

**Comment by tylerjw** [comment]
Unrelated to my comments on how this should be implemented.  It is still failing in CI.  I'm assuming you don't intend for this test to fail in CI.

File: moveit_ros/hybrid_planning/local_planner/local_planner_component/src/local_planner_component.cpp
Comment (tylerjw) [reply moveit_ros/hybrid_planning/local_planner/local_planner_component/src/local_planner_component.cpp:164]: Even lock-free synchronization primitives still cause points of synchronization (single-threaded sections of the code).  I would much prefer a method of communication that does not rely on synchronization of memory access between threads.

The primary reason for this is that if you make something an atomic it has to be stored in ram and access to that point in ram has to be synchronized.  This means that for the CPU to modify or read the value it has to go and get it from ram and then make sure it changed it while preventing no other CPU from accessing it at the same time.  lock-free things like atomic use built-in CPU instructions for this instead of mutexes so they are more performant but they are still creating points in the code that have to be executed in a single-threaded way.

Comment (tylerjw) [reply moveit_ros/hybrid_planning/local_planner/local_planner_component/src/local_planner_component.cpp:164]: My proposal would be to change our mental model for how we build multi-threaded applications to one that does not use mutable state shared between threads for communication.

**Comment by AndyZe** [comment]
Sorry about that. It was passing at one point but seems to be failing due to a ros2_control change:

`[ros2_control_node-5] [WARN] [1642614151.127346724] [fake_generic_system]: The usage of initial_position has been deprecated. Please use 'initial_value' instead.`

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h
Comment (tylerjw) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h:110]: To make it clear from our discussion today I want to find a better way to communicate between threads than raw synchronization primitives around memory.  That however is a much larger discussion and requires more research than this PR.  Please just make sure you are using some mechanism (atomic, mutex, etc) to synchronize access to any variables you are sharing between two concurrent sections.

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h
Comment (tylerjw) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h]: I don't know what the right answer is in this case.  Another question I have is are these callbacks re-entrant and should we guard against that too.  Another way to ask that question is what would happen if on two separate threads ROS calls one of these callbacks twice at the same time?

For now, go ahead and use some sort of synchronization.  Long term we need a better model for how to approach this than shared memory if we want MoveIt to scale onto many threads well.

**Comment by tylerjw** [comment]
> Sorry about that. It was passing at one point but seems to be failing due to a ros2_control change

You don't need to be sorry, we just need to figure this out before we can merge it.  If I have time I'll look into it.

File: moveit_ros/hybrid_planning/test/launch/test_basic_integration.test.py
Comment (tylerjw) [inline moveit_ros/hybrid_planning/test/launch/test_basic_integration.test.py:24]: I wonder if Jafar's change would be useful here?

**Review by tylerjw — APPROVED** [review]
I read through these changes again and it all looks good.  We just need to figure out the ros control issue.

**Comment by AndyZe** [comment]
This moveit_resources PR should fix it:

https://github.com/ros-planning/moveit_resources/pull/110

**Comment by AndyZe** [comment]
After fixing some other issues, I think CI is failing now because we're using your branch of moveit_resources (this commit:  https://github.com/ros-planning/moveit2/commit/8a8249c681a871e4c1e27deb6480418f957cfdb2).

What do you think, @JafarAbdi? Can you at least rebase your fork on the main branch:  https://github.com/ros-planning/moveit_resources/tree/ros2

**Comment by tylerjw** [comment]
I merged the PR of jafar's to fix that.  It looks like we are using the ros2 branch of moveit_resources now: https://github.com/ros-planning/moveit2/blob/b0e2cd0272ae2d99ffb6fd4df0f592caf930c327/moveit2.repos#L12

**Comment by AndyZe** [comment]
Commit 2cc8146  here fixes the warnings in that CI run of main that you triggered, @tylerjw. But there is still another issue after that.

**Comment by AndyZe** [comment]
Started a new PR here to focus on this issue:

https://github.com/ros-planning/moveit2/pull/1025

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp
Comment (henningkayser) [inline moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp:159]: Just saw this now... Shouldn't this use the API of the ServerGoalHandle to flag the goal as canceled, so that the affected code can use `ServerGoalHandle::is_canceling()`? https://docs.ros2.org/latest/api/rclcpp_action/classrclcpp__action_1_1ServerGoalHandle.html

**Review by henningkayser — APPROVED** [review]
I think this is a very good addition, especially with the tests. I haven't tested this locally, but the logic seems very reasonable to me. I'm just not sure if we are fully using the rclcpp_action API as intended.

File: moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp
Comment (AndyZe) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp:159]: If I understand, you would like me to add something like this, here in the cancel callback:

        `hybrid_planning_goal_handle->canceled();`
        
Looking at the [tutorial](https://docs.ros.org/en/foxy/Tutorials/Actions/Writing-a-Cpp-Action-Server-Client.html), they do not use canceled() within the cancel callback. My guess is, `goal->canceled()` would be redundant when used like this.

Comment (AndyZe) [reply moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_manager.cpp:159]: I can follow up with a new PR if I missed something here.

**Pull request merged**
